# Assigment 7
This assigment will be graded if everything works well. I will run the script as once and everything should be done without errors and mistakes. I should be able to run your scripts in my computer and get all the results. USE RELATIVE PATHS. An error or exception or anything that breaks the code will means NO GRADE (0). Additionally, you are not able to modify any file handly. It also means NO GRADE (0). Comment everything you think will help others read your script. We expect 0 errors using GitHub. Everything will be graded!

ASK EVERYTHING! WE ARE HERE TO HELP YOU!

GET YOUR GOOGLE API AND TOKEN. YOU WILL NEED THEM TO DO THIS TASK.

- Import Data from this url. This dataset is in PDF format. You have to convert to PandasDataFrame and keep only the BBVA offices in LIMA.
- Use GoogleMaps API and geocode all the BBVA offices. For those offices that Google API gets no information, use internet and get the latitude and longitude handly and add them to dataset.
- Use Google API to find the driving time (best guess) from all the group members' address and all the LIMA BBVA offices.
- Finally, you have to give a report which offices are the most closest and furthest to every group member's address.

In [2]:
# Packages
import pandas as pd
import os
import urllib.request, json, csv
import numpy as np
from tqdm import tqdm_notebook as tqdm
# For sending GET requests from the API
import requests
# For saving access tokens and for file management when creating and adding to the dataset
import os
# For dealing with json responses we receive from the API
import json
# For displaying the data after
import pandas as pd
# For saving the response data in CSV format
import csv
# For parsing the dates received from twitter in readable formats
import datetime
import dateutil.parser
import unicodedata
#To add wait time between requests
import time
import requests
#maps
import googlemaps

In [3]:
#pip install googlemaps

## 1. First we import the pdf file into pandas and only keep desired rows

Import Data from this url. This dataset is in PDF format. You have to convert to PandasDataFrame and keep only the BBVA offices in LIMA.

Yo lo convertí a excel por fuera y lo subí directo de mis documentos.
Faltaría:
- Encontrar cómo subirlo de PDF para que identifique las tablas
- Usar relative paths

In [4]:
file_path = "Oficinas-BBVA-abiertas-23.10.20.xlsx"

df = pd.read_excel(file_path)

In [5]:
#Convertir la primera fila en el nombre de las columnas

df.rename(columns=df.iloc[0], inplace = True)
df.drop(df.index[0], inplace = True)
df

OFICINA  \
1               BAGUA GRANDE   
2                   CHIMBOTE   
3                     HUARAZ   
4         MEGAPLAZA CHIMBOTE   
5    MERCADO MODELO CHIMBOTE   
..                       ...   
302                    TACNA   
303                   TUMBES   
304               7 DE JUNIO   
305      OPEN PLAZA PUCALLPA   
306                 PUCALLPA   

                                             DIRECCION      DISTRITO  \
1                                 AV. CHACHAPOYAS 2044  BAGUA GRANDE   
2               AV. FRANCISCO BOLOGNESI 696 - CHIMBOTE      CHIMBOTE   
3         CALLE 2 ESQ. JR. JOSE SUCRE - PLAZA DE ARMAS        HUARAZ   
4    AV. VICTOR R. HAYA DE LA TORRE 4694, LC B4 Y B...      CHIMBOTE   
5    JR. ELIAS AGUIRRE 498, ESQ. CON JR. LADISLAO E...      CHIMBOTE   
..                                                 ...           ...   
302                         AV. SAN MARTIN 665 - TACNA         TACNA   
303                                    JR. BOLIVAR 237        TUMBES   
304                           JIRÓN 7 DE JUNIO N.- 201      CALLERIA   
305  AV. CENTENARIO 2086 Y 2192, JR. LAUREANO DEL Á...   YARINACOCHA   
306                     ESQUINA UCAYALI Y RAYMONDI 699      CALLERIA   

             PROVINCIA DEPARTAMENTO  
1            UTCUBAMBA     AMAZONAS  
2                SANTA       ANCASH  
3               HUARAZ       ANCASH  
4                SANTA       ANCASH  
5                SANTA       ANCASH  
..                 ...          ...  
302              TACNA        TACNA  
303             TUMBES       TUMBES  
304  CORONEL\nPORTILLO      UCAYALI  
305  CORONEL\nPORTILLO      UCAYALI  
306   CORONEL PORTILLO      UCAYALI  

[306 rows x 5 columns]

In [6]:
#nos quedamos solo con el departamento de Lima

df = df.loc[df['DEPARTAMENTO'] == "LIMA"]
df = df.reset_index(drop = True)
df

OFICINA  \
0               CAMINO REAL   
1               28 DE JULIO   
2          ABANCAY-CONGRESO   
3                AEROPUERTO   
4              ANGAMOS ESTE   
..                      ...   
201     NESTLE-DONOFRIO 164   
202  BIENESTAR DE LA MARINA   
203                 SEDAPAL   
204         ESPECIAL SENATI   
205          VILLA INFANTAS   

                                             DIRECCION       DISTRITO  \
0                                  AV. CAMINO REAL 355     SAN ISIDRO   
1                                  AV. 28 DE JULIO 775           LIMA   
2                                  AV. ABANCAY 260-262           LIMA   
3         CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A         CALLAO   
4    Av. Angamos Este 215 - 227, esquina con Av. Pe...     MIRAFLORES   
..                                                 ...            ...   
201                                 AV. VENEZUELA 2580           LIMA   
202                          AV. VENEZUELA S/N CDRA 34     SAN MIGUEL   
203  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...    EL AGUSTINO   
204                          AV. ALFREDO MENDIOLA 3520  INDEPENDENCIA   
205  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...     LOS OLIVOS   

    PROVINCIA DEPARTAMENTO  
0        LIMA         LIMA  
1        LIMA         LIMA  
2        LIMA         LIMA  
3        LIMA         LIMA  
4        LIMA         LIMA  
..        ...          ...  
201      LIMA         LIMA  
202      LIMA         LIMA  
203      LIMA         LIMA  
204      LIMA         LIMA  
205      LIMA         LIMA  

[206 rows x 5 columns]

## 2. Then we find the coordinates for each BBVA office

Use GoogleMaps API and geocode all the BBVA offices. For those offices that Google API gets no information, use internet and get the latitude and longitude handly and add them to dataset.


In [7]:
df['direc'] = df['DIRECCION'] + " " + df['DISTRITO'] + " " + "BBVA"
df
#By adding strings from "DISTRITO", there is more information in the adress and we reduce the NANs from 39 to 7.
#Then, by adding "BBVA" to the string we reduce the missing values from 6 to 3.

OFICINA  \
0               CAMINO REAL   
1               28 DE JULIO   
2          ABANCAY-CONGRESO   
3                AEROPUERTO   
4              ANGAMOS ESTE   
..                      ...   
201     NESTLE-DONOFRIO 164   
202  BIENESTAR DE LA MARINA   
203                 SEDAPAL   
204         ESPECIAL SENATI   
205          VILLA INFANTAS   

                                             DIRECCION       DISTRITO  \
0                                  AV. CAMINO REAL 355     SAN ISIDRO   
1                                  AV. 28 DE JULIO 775           LIMA   
2                                  AV. ABANCAY 260-262           LIMA   
3         CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A         CALLAO   
4    Av. Angamos Este 215 - 227, esquina con Av. Pe...     MIRAFLORES   
..                                                 ...            ...   
201                                 AV. VENEZUELA 2580           LIMA   
202                          AV. VENEZUELA S/N CDRA 34     SAN MIGUEL   
203  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...    EL AGUSTINO   
204                          AV. ALFREDO MENDIOLA 3520  INDEPENDENCIA   
205  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...     LOS OLIVOS   

    PROVINCIA DEPARTAMENTO                                              direc  
0        LIMA         LIMA                AV. CAMINO REAL 355 SAN ISIDRO BBVA  
1        LIMA         LIMA                      AV. 28 DE JULIO 775 LIMA BBVA  
2        LIMA         LIMA                      AV. ABANCAY 260-262 LIMA BBVA  
3        LIMA         LIMA  CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A C...  
4        LIMA         LIMA  Av. Angamos Este 215 - 227, esquina con Av. Pe...  
..        ...          ...                                                ...  
201      LIMA         LIMA                       AV. VENEZUELA 2580 LIMA BBVA  
202      LIMA         LIMA          AV. VENEZUELA S/N CDRA 34 SAN MIGUEL BBVA  
203      LIMA         LIMA  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...  
204      LIMA         LIMA       AV. ALFREDO MENDIOLA 3520 INDEPENDENCIA BBVA  
205      LIMA         LIMA  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...  

[206 rows x 6 columns]

In [8]:
direc = df["direc"]
direc.head()

0                  AV. CAMINO REAL 355 SAN ISIDRO BBVA
1                        AV. 28 DE JULIO 775 LIMA BBVA
2                        AV. ABANCAY 260-262 LIMA BBVA
3    CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A C...
4    Av. Angamos Este 215 - 227, esquina con Av. Pe...
Name: direc, dtype: object

In [9]:
coord = np.zeros(shape=( len(direc), 2), dtype =float)
i=0

gmaps = googlemaps.Client(key='AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g')

# Loop over lists
for d in tqdm(direc):   

    # Geocoding an address
    geocode_result = gmaps.geocode( d , region = 'pe')
    
    # Check the len of result
    if len(geocode_result)==0 :
        coord[i][0] = np.nan
        coord[i][1] = np.nan

    
    # Get info
    else :
        coord[i][0] = geocode_result[0]['geometry']['location']['lat']
        coord[i][1] = geocode_result[0]['geometry']['location']['lng']
        
    i=i+1

<ipython-input-9-890b219f9aea>:7: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  for d in tqdm(direc):


  0%|          | 0/206 [00:00<?, ?it/s]

In [10]:
#Create the dataframe with the new coordenates
coords = pd.DataFrame( coord.tolist() , columns = ["Lat", "Lon"] )
coords['direc'] = direc
coords

Lat        Lon                                              direc
0   -12.097362 -77.036099                AV. CAMINO REAL 355 SAN ISIDRO BBVA
1   -12.064030 -77.037961                      AV. 28 DE JULIO 775 LIMA BBVA
2   -12.047955 -77.026997                      AV. ABANCAY 260-262 LIMA BBVA
3   -12.023538 -77.104397  CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A C...
4   -12.113476 -77.028582  Av. Angamos Este 215 - 227, esquina con Av. Pe...
..         ...        ...                                                ...
201 -12.062448 -77.093581                       AV. VENEZUELA 2580 LIMA BBVA
202 -12.062026 -77.092444          AV. VENEZUELA S/N CDRA 34 SAN MIGUEL BBVA
203 -12.033125 -76.983811  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...
204 -11.994069 -77.063041       AV. ALFREDO MENDIOLA 3520 INDEPENDENCIA BBVA
205 -11.959300 -77.076025  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...

[206 rows x 3 columns]

In [11]:
#Now we merge the coordenates dataframe with the original dataframe into the final df
df_final = df.merge(coords, left_on='direc', right_on='direc', how='left')
df_final.drop('direc', inplace=True, axis=1)
df_final.index.name = 'id'
df_final.reset_index(inplace=True)
df_final

id                 OFICINA  \
0      0             CAMINO REAL   
1      1             28 DE JULIO   
2      2        ABANCAY-CONGRESO   
3      3              AEROPUERTO   
4      4            ANGAMOS ESTE   
..   ...                     ...   
201  201     NESTLE-DONOFRIO 164   
202  202  BIENESTAR DE LA MARINA   
203  203                 SEDAPAL   
204  204         ESPECIAL SENATI   
205  205          VILLA INFANTAS   

                                             DIRECCION       DISTRITO  \
0                                  AV. CAMINO REAL 355     SAN ISIDRO   
1                                  AV. 28 DE JULIO 775           LIMA   
2                                  AV. ABANCAY 260-262           LIMA   
3         CENTRO AÉREO COMERCIAL LOCALES 110 A Y 111 A         CALLAO   
4    Av. Angamos Este 215 - 227, esquina con Av. Pe...     MIRAFLORES   
..                                                 ...            ...   
201                                 AV. VENEZUELA 2580           LIMA   
202                          AV. VENEZUELA S/N CDRA 34     SAN MIGUEL   
203  KM. 1 AUTOPISTA RAMIRO PRIALE (AL INTERIOR DE ...    EL AGUSTINO   
204                          AV. ALFREDO MENDIOLA 3520  INDEPENDENCIA   
205  MZ. K, SUB LOTE 4A, PARCELA A-1. LOTIZACIÓN IN...     LOS OLIVOS   

    PROVINCIA DEPARTAMENTO        Lat        Lon  
0        LIMA         LIMA -12.097362 -77.036099  
1        LIMA         LIMA -12.064030 -77.037961  
2        LIMA         LIMA -12.047955 -77.026997  
3        LIMA         LIMA -12.023538 -77.104397  
4        LIMA         LIMA -12.113476 -77.028582  
..        ...          ...        ...        ...  
201      LIMA         LIMA -12.062448 -77.093581  
202      LIMA         LIMA -12.062026 -77.092444  
203      LIMA         LIMA -12.033125 -76.983811  
204      LIMA         LIMA -11.994069 -77.063041  
205      LIMA         LIMA -11.959300 -77.076025  

[206 rows x 8 columns]

### Now we replace NAs by hand

In [12]:
isnan = df_final[df_final.isna().any(axis=1)]
len(isnan)
#There are 3 cases with missing values: Mala, Naciones Unidas & Strip Center Barranca

2

In [13]:
isnan

id OFICINA        DIRECCION     DISTRITO PROVINCIA DEPARTAMENTO  Lat  \
40    40   CAPON  JR. UCAYALI 735  LA VICTORIA      LIMA         LIMA  NaN   
113  113    MALA     JR. REAL 305         MALA    CAÑETE         LIMA  NaN   

     Lon  
40   NaN  
113  NaN

In [14]:
#Oficina Mala
df_final.loc[113,'Lat']= -12.6558538
df_final.loc[113,'Lon']= -76.6348772
#Oficina Naciones Unidas
df_final.loc[127,'Lat']= -12.1003510
df_final.loc[127,'Lon']= -77.0636233
#Oficina Strip Center Barranca
df_final.loc[40,'Lat']= -10.75237060
df_final.loc[40,'Lon']= -77.75986317

## 3. Find driving time

Use Google API to find the driving time (best guess) from all the group members' address and all the LIMA BBVA offices.

Les puse direcciones al azar. Sería cuestión que lo editen.

#### Group members' address
- Alejandro: -12.1148761,-77.0386600
- Steph: -12.1148389,-77.0117565
- Juan Carlos: -12.0961813,-77.0714548
- Gabriela: -12.0259368,-77.0069487

In [15]:
#We subset only coordenates and transform them to strings instead of floats
bbva = df_final.iloc[:, [6,7]].applymap(str)

In [16]:
#Create a list of "destinations"
bbva['destino'] = bbva['Lat'] + "," + bbva['Lon']
destino = bbva['destino'].tolist()
destino

['-12.0973616,-77.0360994',
 '-12.0640305,-77.0379609',
 '-12.0479547,-77.02699679999999',
 '-12.0235384,-77.10439749999999',
 '-12.1134759,-77.028582',
 '-12.1120524,-77.0117853',
 '-12.0719308,-77.011235',
 '-12.1036288,-77.0307414',
 '-12.0716113,-77.0374212',
 '-12.0915794,-77.03373189999999',
 '-12.1140109,-77.0480248',
 '-12.0307591,-77.0848436',
 '-12.0807507,-77.0665087',
 '-12.0965008,-77.0020389',
 '-12.0470329,-76.93830129999999',
 '-10.752491,-77.7598701',
 '-12.1473578,-77.02133959999999',
 '-12.0370867,-77.1372552',
 '-12.0534252,-77.0958897',
 '-12.1266464,-77.01623',
 '-12.0722545,-77.0661447',
 '-12.0358864,-77.0933675',
 '-12.1217533,-77.0117212',
 '-12.0622538,-77.0665705',
 '-12.0088873,-77.0602449',
 '-12.1285546,-76.9993054',
 '-12.104926,-76.9707944',
 '-12.12913,-76.9841241',
 '-12.0833319,-77.03506999999999',
 '-12.0954325,-77.02623779999999',
 '-12.076778,-77.083096',
 '-12.0754679,-77.064285',
 '-12.0608102,-77.143934',
 '-12.0447042,-76.9343444',
 '-12.04811

In [17]:
#Alejandro
df_driving = pd.DataFrame (destino, columns = ['destination'])
df_driving['origin'] = "-12.1148761,-77.0386600"
df_driving

destination                   origin
0           -12.0973616,-77.0360994  -12.1148761,-77.0386600
1           -12.0640305,-77.0379609  -12.1148761,-77.0386600
2    -12.0479547,-77.02699679999999  -12.1148761,-77.0386600
3    -12.0235384,-77.10439749999999  -12.1148761,-77.0386600
4            -12.1134759,-77.028582  -12.1148761,-77.0386600
..                              ...                      ...
201         -12.0624481,-77.0935808  -12.1148761,-77.0386600
202         -12.0620255,-77.0924442  -12.1148761,-77.0386600
203         -12.0331248,-76.9838113  -12.1148761,-77.0386600
204         -11.9940687,-77.0630414  -12.1148761,-77.0386600
205  -11.9593005,-77.07602539999999  -12.1148761,-77.0386600

[206 rows x 2 columns]

In [18]:
#Steph
df_driving2 = pd.DataFrame (destino, columns = ['destination'])
df_driving2['origin'] = "-12.1148389,-77.0117565"
df_driving2

destination                   origin
0           -12.0973616,-77.0360994  -12.1148389,-77.0117565
1           -12.0640305,-77.0379609  -12.1148389,-77.0117565
2    -12.0479547,-77.02699679999999  -12.1148389,-77.0117565
3    -12.0235384,-77.10439749999999  -12.1148389,-77.0117565
4            -12.1134759,-77.028582  -12.1148389,-77.0117565
..                              ...                      ...
201         -12.0624481,-77.0935808  -12.1148389,-77.0117565
202         -12.0620255,-77.0924442  -12.1148389,-77.0117565
203         -12.0331248,-76.9838113  -12.1148389,-77.0117565
204         -11.9940687,-77.0630414  -12.1148389,-77.0117565
205  -11.9593005,-77.07602539999999  -12.1148389,-77.0117565

[206 rows x 2 columns]

In [19]:
#Gabriela
df_driving3 = pd.DataFrame (destino, columns = ['destination'])
df_driving3['origin'] = "-12.0259368,-77.0069487"
df_driving3

destination                   origin
0           -12.0973616,-77.0360994  -12.0259368,-77.0069487
1           -12.0640305,-77.0379609  -12.0259368,-77.0069487
2    -12.0479547,-77.02699679999999  -12.0259368,-77.0069487
3    -12.0235384,-77.10439749999999  -12.0259368,-77.0069487
4            -12.1134759,-77.028582  -12.0259368,-77.0069487
..                              ...                      ...
201         -12.0624481,-77.0935808  -12.0259368,-77.0069487
202         -12.0620255,-77.0924442  -12.0259368,-77.0069487
203         -12.0331248,-76.9838113  -12.0259368,-77.0069487
204         -11.9940687,-77.0630414  -12.0259368,-77.0069487
205  -11.9593005,-77.07602539999999  -12.0259368,-77.0069487

[206 rows x 2 columns]

In [33]:
#Juan Carlos
df_driving4 = pd.DataFrame (destino, columns = ['destination'])
df_driving4['origin'] = "-12.0961813,-77.0714548"
df_driving4

destination                   origin
0           -12.0973616,-77.0360994  -12.0961813,-77.0714548
1           -12.0640305,-77.0379609  -12.0961813,-77.0714548
2    -12.0479547,-77.02699679999999  -12.0961813,-77.0714548
3    -12.0235384,-77.10439749999999  -12.0961813,-77.0714548
4            -12.1134759,-77.028582  -12.0961813,-77.0714548
..                              ...                      ...
201         -12.0624481,-77.0935808  -12.0961813,-77.0714548
202         -12.0620255,-77.0924442  -12.0961813,-77.0714548
203         -12.0331248,-76.9838113  -12.0961813,-77.0714548
204         -11.9940687,-77.0630414  -12.0961813,-77.0714548
205  -11.9593005,-77.07602539999999  -12.0961813,-77.0714548

[206 rows x 2 columns]

In [34]:
df_driving_final = pd.concat([df_driving, df_driving2, df_driving3, df_driving4]).reset_index(drop=True)
df_driving_final

destination                   origin
0           -12.0973616,-77.0360994  -12.1148761,-77.0386600
1           -12.0640305,-77.0379609  -12.1148761,-77.0386600
2    -12.0479547,-77.02699679999999  -12.1148761,-77.0386600
3    -12.0235384,-77.10439749999999  -12.1148761,-77.0386600
4            -12.1134759,-77.028582  -12.1148761,-77.0386600
..                              ...                      ...
819         -12.0624481,-77.0935808  -12.0961813,-77.0714548
820         -12.0620255,-77.0924442  -12.0961813,-77.0714548
821         -12.0331248,-76.9838113  -12.0961813,-77.0714548
822         -11.9940687,-77.0630414  -12.0961813,-77.0714548
823  -11.9593005,-77.07602539999999  -12.0961813,-77.0714548

[824 rows x 2 columns]

In [35]:
df_driving_final['combination'] = list(range(1,len(df_driving_final)+1))
df_driving_final

destination                   origin  combination
0           -12.0973616,-77.0360994  -12.1148761,-77.0386600            1
1           -12.0640305,-77.0379609  -12.1148761,-77.0386600            2
2    -12.0479547,-77.02699679999999  -12.1148761,-77.0386600            3
3    -12.0235384,-77.10439749999999  -12.1148761,-77.0386600            4
4            -12.1134759,-77.028582  -12.1148761,-77.0386600            5
..                              ...                      ...          ...
819         -12.0624481,-77.0935808  -12.0961813,-77.0714548          820
820         -12.0620255,-77.0924442  -12.0961813,-77.0714548          821
821         -12.0331248,-76.9838113  -12.0961813,-77.0714548          822
822         -11.9940687,-77.0630414  -12.0961813,-77.0714548          823
823  -11.9593005,-77.07602539999999  -12.0961813,-77.0714548          824

[824 rows x 3 columns]

In [36]:
#Create the three lists: combination, origin and destination
combination = df_driving_final['combination'].tolist()
origin = df_driving_final['origin'].tolist()
destination = df_driving_final['destination'].tolist()

In [37]:
data_distance = {} 

In [38]:
# Loop to generate info about geolocations
duration_info = np.zeros(shape=(len(combination),2), dtype =float)
i=0

for c,o,d in tqdm(list(zip(combination, origin, destination))):
    try:
        # Google MapsDdirections API endpoint
        endpoint = 'https://maps.googleapis.com/maps/api/directions/json?'

    ## Fixed Parameters
        # Paramaters
        traffic_model = 'best_guess'  

        # Departure time
        departure_time = 'now'

        # driving, walking, biclycling, transit
        mode = 'driving'

        # key
        api_key = 'AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g'

        # region to look for (spain= es , germany = de , Switzerland= swiss)
        region = 'pe'

     ## Parameters
        # Origin
        origin = o
        
        # Destinations
        destination = d
        
        #Building the URL for the request
        nav_request = 'origin={}&destination={}&departure_time={}&traffic_model={}&mode={}&region={}&key={}'.format(origin , 
                        destination , departure_time , traffic_model , mode, region, api_key)
        
        # Concatenate strings
        request = endpoint + nav_request
        print(request)

        #Sends the request and reads the response.
        response = urllib.request.urlopen(request).read()
        
        

        #Loads response as JSON
        directions = json.loads(response)

        legs = directions['routes'][0]['legs'][0]
        #print(legs)
        
        duration_info[i][0] = c
        duration_info[i][1] = legs['duration']['value']
        #print(duration_info[i][1])
        
        i=i+1

        my_keys = ['duration']
        info = { my_key: legs[my_key] for my_key in my_keys }
        data_distance[c] = info 
    
    except Exception as e:
        
        duration_info[i][0] = c
        duration_info[i][1] = "nan"
        
        i=i+1
        
        print(c)

<ipython-input-38-9f36dc2d5413>:5: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  for c,o,d in tqdm(list(zip(combination, origin, destination))):


  0%|          | 0/824 [00:00<?, ?it/s]

https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-12.0973616,-77.0360994&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-12.0640305,-77.0379609&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-12.0479547,-77.02699679999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-12.1148761,-77.0386600&destination=-12.0235384,-77.10439749999999&departure_time=now&traffic_model=best_guess&mode=driving&region=pe&key=AIzaSyD5CNJfBkf_OIH_D4shwz_Bz8MRgoJBe_g
https://maps.googleapis.com/maps/api/directions/json?origin=-1

In [39]:
duration_info

array([[1.000e+00, 4.730e+02],
       [2.000e+00, 1.266e+03],
       [3.000e+00, 2.083e+03],
       ...,
       [8.220e+02, 2.395e+03],
       [8.230e+02, 2.685e+03],
       [8.240e+02, 3.081e+03]])

In [40]:
distance_api = pd.DataFrame(duration_info, columns = ['Combination', 'Duration_seg'])
distance_api

Combination  Duration_seg
0            1.0         473.0
1            2.0        1266.0
2            3.0        2083.0
3            4.0        2131.0
4            5.0         401.0
..           ...           ...
819        820.0        1337.0
820        821.0        1360.0
821        822.0        2395.0
822        823.0        2685.0
823        824.0        3081.0

[824 rows x 2 columns]

## 4. Report offices closets and furthest from each group members' adress

Se me ocurre que podríamos crear un dataframe en base a lo anterior (cuando se arregle) que sea una columna por miembro del grupo con el tiempo a cada una de las 205 oficinas, y hacemos un merge con la df_final. Luego que nos dé la fila donde está el valor mínimo y máximo para cada una de las 4 columnas y así identificamos las oficinas.

In [41]:
group_size = 206  # Set group size to 206
num_groups = (len(distance_api) - 1) // group_size + 1  # Calculate number of groups needed
distance_api['group'] = [i // group_size + 1 for i in range(len(distance_api))]  # Assign group numbers


In [42]:
distance_api

Combination  Duration_seg  group
0            1.0         473.0      1
1            2.0        1266.0      1
2            3.0        2083.0      1
3            4.0        2131.0      1
4            5.0         401.0      1
..           ...           ...    ...
819        820.0        1337.0      4
820        821.0        1360.0      4
821        822.0        2395.0      4
822        823.0        2685.0      4
823        824.0        3081.0      4

[824 rows x 3 columns]

In [43]:
#Alejandro
distance_api_1 = distance_api[distance_api["group"] == 1]
distance_api_1.sort_values(by=['Duration_seg'])

Combination  Duration_seg  group
174        175.0         217.0      1
138        139.0         242.0      1
55          56.0         264.0      1
53          54.0         301.0      1
42          43.0         314.0      1
..           ...           ...    ...
15          16.0       12464.0      1
51          52.0       27252.0      1
158        159.0      105224.0      1
92          93.0      184095.0      1
148        149.0           NaN      1

[206 rows x 3 columns]

The BBVAs closest to Alejandro's house are those belonging to the combination with the 175, 139 and 56 BBVA in the list. While the farthest are those belonging to the combination with the 149, 93 and 159 BBVA in the list. The nearest is the Plaza Union office, and the fartest is the Santa Monica office. 

In [50]:
df_list = df[(df.index == 175) | (df.index == 149)]
df_list

OFICINA             DIRECCION    DISTRITO PROVINCIA DEPARTAMENTO  \
149   PLAZA UNIÓN  AV. ARGENTINA N. 500        LIMA      LIMA         LIMA   
175  SANTA MONICA   AV. JUAN PEZET 1425  SAN ISIDRO      LIMA         LIMA   

                                   direc  
149       AV. ARGENTINA N. 500 LIMA BBVA  
175  AV. JUAN PEZET 1425 SAN ISIDRO BBVA

In [44]:
#Steph
distance_api_2 = distance_api[distance_api["group"] == 2]
distance_api_2.sort_values(by=['Duration_seg'])

Combination  Duration_seg  group
228        229.0         268.0      2
211        212.0         310.0      2
225        226.0         384.0      2
385        386.0         389.0      2
349        350.0         418.0      2
..           ...           ...    ...
221        222.0       12960.0      2
257        258.0       27108.0      2
364        365.0      105263.0      2
298        299.0      183404.0      2
354        355.0           NaN      2

[206 rows x 3 columns]

The BBVAs closest to Stephy's house are those belonging to the combination with the 24, 7 and 21 BBVA in the list (229, 212 and 226). While the farthest are those belonging to the combination with the 150, 94 and 160 BBVA in the list (355, 299 and 365). The nearest is the C.C Fiori office and the fartest is the Parque Industrial in Villa El Salvador office. 

In [51]:
df_list_2 = df[(df.index == 24) | (df.index == 150)]
df_list_2

OFICINA  \
24                     C.C. FIORI   
150  PQUE. IND. VILLA EL SALVADOR   

                                             DIRECCION              DISTRITO  \
24                         AV. TOMÁS VALLE 499G, LC 14  SAN MARTIN DE PORRES   
150  MZ. M1, LT. 1 Y 2, PARCELA II, PARQUE INDUSTRI...     VILLA EL SALVADOR   

    PROVINCIA DEPARTAMENTO                                              direc  
24       LIMA         LIMA  AV. TOMÁS VALLE 499G, LC 14 SAN MARTIN DE PORR...  
150      LIMA         LIMA  MZ. M1, LT. 1 Y 2, PARCELA II, PARQUE INDUSTRI...

In [32]:
#Gabriela
distance_api_2 = distance_api[distance_api["group"] == 3]
distance_api_2.sort_values(by=['Duration_seg'])

Combination  Duration_seg  group
539        540.0         243.0      3
524        525.0         262.0      3
497        498.0         336.0      3
543        544.0         385.0      3
587        588.0         510.0      3
..           ...           ...    ...
427        428.0       12256.0      3
463        464.0       27019.0      3
570        571.0      104949.0      3
504        505.0      184245.0      3
560        561.0           NaN      3

[206 rows x 3 columns]

The BBVAs closest to Gabriela's house are those belonging to the combination with the 130, 116 and 88  BBVAs of the lista (540, 526 and 498), while the farthest are those belonging to the combination with 151, 95 and 191 BBVA of the list (561, 505 and 571).The nearest is the Nicolas Ayllon office and the fartest is the Pro office.

In [52]:
df_list_3 = df[(df.index == 130) | (df.index == 151)]
df_list_3

OFICINA                     DIRECCION    DISTRITO PROVINCIA  \
130  NICOLAS AYLLON  AV. NICOLAS AYLLON 2032-2036         ATE      LIMA   
151             PRO     AV. ALFREDO MENDIOLA 7859  LOS OLIVOS      LIMA   

    DEPARTAMENTO                                      direc  
130         LIMA      AV. NICOLAS AYLLON 2032-2036 ATE BBVA  
151         LIMA  AV. ALFREDO MENDIOLA 7859 LOS OLIVOS BBVA

In [45]:
#JC
distance_api_4 = distance_api[distance_api["group"] == 4]
distance_api_4.sort_values(by=['Duration_seg'])

Combination  Duration_seg  group
745        746.0         243.0      4
730        731.0         262.0      4
703        704.0         336.0      4
749        750.0         385.0      4
793        794.0         510.0      4
..           ...           ...    ...
633        634.0       12256.0      4
669        670.0       27576.0      4
776        777.0      104949.0      4
710        711.0      184245.0      4
766        767.0           NaN      4

[206 rows x 3 columns]

The BBVAs closest to Juan Carlos' house are those belonging to the combination with the 131, 116 and 89 BBVAs of the list (746, 731 and 704). While the farthest are those belonging to the combination with the 162, 96 and 152 BBVA of the list (777, 711 and 767). The nearest is the Orrantia office and the fartest is the Samuel del Alcalzar office.

In [54]:
df_list_4 = df[(df.index == 131) | (df.index == 162)]
df_list_4

OFICINA                       DIRECCION           DISTRITO  \
131            ORRANTIA      AV. ALBERTO DEL CAMPO 1006  MAGDALENA DEL MAR   
162  SAMUEL DEL ALCAZAR  AV. AMANCAES N° 192 - 194 -198              RIMAC   

    PROVINCIA DEPARTAMENTO                                              direc  
131      LIMA         LIMA  AV. ALBERTO DEL CAMPO 1006 MAGDALENA DEL MAR BBVA  
162      LIMA         LIMA          AV. AMANCAES N° 192 - 194 -198 RIMAC BBVA